File: web/src/lib/crypto/wrap.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 PRSN EX Inc.
// ECDH-ES+A256KW key wrapping (Envelope §5 DEK wrap + §7.2 metadata-key wrap).
// The composition step: a fresh ephemeral-static ECDH → Concat-KDF-SHA-256 (with
// the P-015 party_v binding) → AES-256 Key Wrap, serialized as the JOSE-shaped
// WrapEnvelope. Mirrors signet-crypto's wrap.rs; it reuses the validated
// ecdh / concatkdf / keywrap primitives and adds only the glue and the
// on-the-wire shape — no new cryptography.
//
// The DEK wrap and the metadata-key wrap are the SAME construction; they differ
// only in party_v: empty for a DEK, the 16-byte root_folder_id for a metadata
// key (P-015). The folder binding means a metadata-key blob moved to another
// folder derives a different KEK and fails AES-KW's integrity check.

import { b64uDecode, b64uEncode, concatBytes, type Bytes } from './bytes';
import { ecdhEsA256kwKek } from './concatkdf';
import { deriveZ, generateEphemeralKeypair, importEcdhPublicX963 } from './ecdh';
import { InvalidInputError } from './errors';
import { aesKwUnwrap, aesKwWrap } from './keywrap';
import { fingerprint } from './pubkey';

export const ECDH_ES_A256KW = 'ECDH-ES+A256KW';

/** The ephemeral public key in JWK form (RFC 7517 §3), as it appears in the
 *  wrap envelope's `epk` field. */
export interface EpkJwk {
  kty: string;
  crv: string;
  /** base64url(no-pad) of the 32-byte X coordinate. */
  x: string;
  /** base64url(no-pad) of the 32-byte Y coordinate. */
  y: string;
}

/** A wrapped-key envelope (Envelope §5 for a DEK, §7.2 for a metadata key —
 *  identical shape). Stored opaquely server-side; the server never parses it. */
export interface WrapEnvelope {
  v: number;
  alg: string;
  epk: EpkJwk;
  /** base64url(no-pad) of the 40-byte AES-KW output. */
  ct: string;
  /** Recipient KEM pubkey fingerprint, lowercase hex (routing only). */
  rfp: string;
}

const EMPTY = new Uint8Array(0);

/** Strictly parse an unknown value as a CLASSICAL wrap envelope: the exact
 *  field set — `ek` or any other extra field is rejected (the §5.2 symmetric
 *  validation / N5: no field-smuggling across algs; the deny_unknown_fields
 *  mirror of the Rust WrapEnvelope). The alg VALUE is checked at unwrap. */
export function parseWrapEnvelope(value: unknown): WrapEnvelope {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new InvalidInputError('wrap envelope must be an object');
  }
  const v = value as Record<string, unknown>;
  if (Object.keys(v).sort().join(',') !== 'alg,ct,epk,rfp,v') {
    throw new InvalidInputError('wrap envelope must have exactly v/alg/epk/ct/rfp');
  }
  if (
    typeof v.v !== 'number' ||
    typeof v.alg !== 'string' ||
    typeof v.ct !== 'string' ||
    typeof v.rfp !== 'string'
  ) {
    throw new InvalidInputError('wrap envelope field types');
  }
  const epk = v.epk;
  if (typeof epk !== 'object' || epk === null || Array.isArray(epk)) {
    throw new InvalidInputError('wrap envelope epk');
  }
  const e = epk as Record<string, unknown>;
  if (Object.keys(e).sort().join(',') !== 'crv,kty,x,y') {
    throw new InvalidInputError('wrap envelope epk must have exactly kty/crv/x/y');
  }
  if (
    typeof e.kty !== 'string' ||
    typeof e.crv !== 'string' ||
    typeof e.x !== 'string' ||
    typeof e.y !== 'string'
  ) {
    throw new InvalidInputError('wrap envelope epk field types');
  }
  return {
    v: v.v,
    alg: v.alg,
    epk: { kty: e.kty, crv: e.crv, x: e.x, y: e.y },
    ct: v.ct,
    rfp: v.rfp,
  };
}

function epkToX963(epk: EpkJwk): Bytes {
  if (epk.kty !== 'EC' || epk.crv !== 'P-256') {
    throw new InvalidInputError('epk must be EC P-256');
  }
  const x = b64uDecode(epk.x);
  const y = b64uDecode(epk.y);
  if (x.length !== 32 || y.length !== 32) {
    throw new InvalidInputError('epk coordinates must be 32 bytes');
  }
  return concatBytes(new Uint8Array([0x04]), x, y);
}

function epkFromX963(x963: Bytes): EpkJwk {
  if (x963.length !== 65 || x963[0] !== 0x04) {
    throw new InvalidInputError('x9.63 uncompressed point');
  }
  return {
    kty: 'EC',
    crv: 'P-256',
    x: b64uEncode(x963.slice(1, 33)),
    y: b64uEncode(x963.slice(33, 65)),
  };
}

// --- unwrap (download / decrypt path) ---

/** Unwrap a DEK wrap with the recipient's ECDH private key (Envelope §5). */
export function unwrapDek(recipientPrivateKey: CryptoKey, envelope: WrapEnvelope): Promise<Bytes> {
  return unwrapWith(recipientPrivateKey, envelope, EMPTY);
}

/** Unwrap a metadata-key wrap. rootFolderId MUST match the folder the wrap was
 *  created for; a mismatch derives a different KEK and AES-KW rejects it (P-015). */
export function unwrapMetadataKey(
  recipientPrivateKey: CryptoKey,
  envelope: WrapEnvelope,
  rootFolderId: Bytes,
): Promise<Bytes> {
  if (rootFolderId.length !== 16) throw new InvalidInputError('root_folder_id must be 16 bytes');
  return unwrapWith(recipientPrivateKey, envelope, rootFolderId);
}

async function unwrapWith(
  recipientPrivateKey: CryptoKey,
  envelope: WrapEnvelope,
  partyV: Bytes,
): Promise<Bytes> {
  if (envelope.alg !== ECDH_ES_A256KW) throw new InvalidInputError('unknown wrap alg');
  const ephemeralPub = await importEcdhPublicX963(epkToX963(envelope.epk));
  const z = await deriveZ(recipientPrivateKey, ephemeralPub);
  const kek = await ecdhEsA256kwKek(z, partyV);
  const key = await aesKwUnwrap(kek, b64uDecode(envelope.ct));
  if (key.length !== 32) throw new InvalidInputError('unwrapped key is not 32 bytes');
  return key;
}

// --- wrap (upload / encrypt path) ---

/** Wrap a 32-byte DEK to a recipient's KEM public key (Envelope §5). No folder
 *  binding (the §4.1 file envelope binds content to file_id via AAD). */
export function wrapDek(recipientPubX963: Bytes, dek: Bytes): Promise<WrapEnvelope> {
  return wrapTo(recipientPubX963, dek, EMPTY);
}

/** Wrap a 32-byte metadata key to a recipient, bound to rootFolderId (P-015). */
export function wrapMetadataKey(
  recipientPubX963: Bytes,
  metadataKey: Bytes,
  rootFolderId: Bytes,
): Promise<WrapEnvelope> {
  if (rootFolderId.length !== 16) throw new InvalidInputError('root_folder_id must be 16 bytes');
  return wrapTo(recipientPubX963, metadataKey, rootFolderId);
}

async function wrapTo(recipientPubX963: Bytes, key: Bytes, partyV: Bytes): Promise<WrapEnvelope> {
  if (key.length !== 32) throw new InvalidInputError('wrapped key must be 32 bytes');
  const recipientPub = await importEcdhPublicX963(recipientPubX963);
  const ephemeral = await generateEphemeralKeypair();
  const z = await deriveZ(ephemeral.privateKey, recipientPub);
  const kek = await ecdhEsA256kwKek(z, partyV);
  const wrapped = await aesKwWrap(kek, key);
  const ephemeralPubX963 = new Uint8Array(
    await crypto.subtle.exportKey('raw', ephemeral.publicKey),
  );
  return {
    v: 1,
    alg: ECDH_ES_A256KW,
    epk: epkFromX963(ephemeralPubX963),
    ct: b64uEncode(wrapped),
    rfp: await fingerprint(recipientPubX963),
  };
}

// --- self-public-key verification (pre-launch crypto review #5) ---

/** Verify that a NON-extractable KEM private key is the counterpart of a claimed
 *  public key — without needing the private key to be extractable. Wraps a random
 *  probe to `publicX963` (ephemeral-static ECDH → Concat-KDF KEK → AES-KW) then
 *  unwraps it with `privateKey`: the round-trip recovers the probe iff the two are
 *  an ECDH pair, because a mismatched public key derives a different KEK and
 *  AES-KW's integrity check fails on unwrap.
 *
 *  This is the only local check available at sign-in, where the self KEM public
 *  key is server-supplied while the private key is recovered NON-extractable from
 *  the PRF-wrapped blob (so WebCrypto cannot re-derive its public key for a direct
 *  comparison). It defends the self-wrap path against a malicious server returning
 *  a substituted self-public-key (which would otherwise wrap the user's own new
 *  uploads to the attacker). Fails closed: any error (malformed key, KEK mismatch)
 *  returns false. */
export async function kemKeypairMatches(
  privateKey: CryptoKey,
  publicX963: Bytes,
): Promise<boolean> {
  const probe = crypto.getRandomValues(new Uint8Array(32)) as Bytes;
  try {
    const envelope = await wrapDek(publicX963, probe);
    const recovered = await unwrapDek(privateKey, envelope);
    return recovered.length === probe.length && recovered.every((b, i) => b === probe[i]);
  } catch {
    return false;
  }
}

File: crypto/src/wrap.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 PRSN EX Inc.

//! `ECDH-ES+A256KW` key wrapping (Envelope §5 DEK wrap + §7.2 metadata-key wrap).
//!
//! Wraps a 32-byte key — a file DEK, or a per-folder metadata key — to a
//! recipient's ECDH P-256 KEM public key: a fresh ephemeral-static ECDH →
//! Concat-KDF-SHA-256 → AES-256 Key Wrap, serialized as the JOSE-shaped JSON
//! envelope. This is the **composition** step (the genuinely novel surface):
//! it reuses the validated [`ecdh`], [`concatkdf`], and [`keywrap`] primitives
//! and adds only the glue and the on-the-wire shape — no new cryptography.
//!
//! The DEK wrap and the metadata-key wrap are the *same* construction; they
//! differ only in the Concat-KDF `PartyVInfo`: empty for a DEK, the
//! `root_folder_id` for a metadata key (P-015). The folder binding means a
//! metadata-key blob moved to another folder derives a different KEK and fails
//! AES-KW's integrity check (Envelope §7.2).

use base64::Engine;
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use p256::elliptic_curve::sec1::ToEncodedPoint;
use p256::{PublicKey, SecretKey};
use rand_core::OsRng;
use serde::{Deserialize, Serialize};
use zeroize::{Zeroize, Zeroizing};

use crate::alg::AlgId;
use crate::error::{CryptoError, Result};
use crate::{concatkdf, ecdh, keywrap, pubkey};

/// The ephemeral public key in JWK form (RFC 7517 §3), as it appears in the
/// wrap envelope's `epk` field.
///
/// `deny_unknown_fields`: both envelope emitters (this crate and the web
/// client) produce exactly `kty`/`crv`/`x`/`y`, so extra JWK members (`kid`,
/// `use`, …) are rejected rather than silently carried — the same strict-parse
/// posture as the envelopes themselves (PQR Spec §5.2).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct EpkJwk {
    pub kty: String,
    pub crv: String,
    /// base64url(no-pad) of the 32-byte X coordinate.
    pub x: String,
    /// base64url(no-pad) of the 32-byte Y coordinate.
    pub y: String,
}

/// A wrapped-key envelope (Envelope §5 for a DEK, §7.2 for a metadata key —
/// identical shape). Stored opaquely server-side in `file_recipients.wrapped_dek`
/// / `wrapped_metadata_keys.wrapped_key`; the server never parses it.
///
/// `deny_unknown_fields` enforces the classical side of the PQR-Spec §5.2
/// symmetric field validation: a classical `ECDH-ES+A256KW` envelope MUST NOT
/// carry a hybrid `ek`/`wk` — an injected field is rejected at parse time
/// (negative test N5), so no field-smuggling across algs.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WrapEnvelope {
    pub v: u32,
    pub alg: String,
    pub epk: EpkJwk,
    /// base64url(no-pad) of the 40-byte AES-KW output.
    pub ct: String,
    /// Recipient KEM pubkey fingerprint, lowercase hex (routing only).
    pub rfp: String,
}

impl WrapEnvelope {
    /// The ephemeral public key from `epk`, as a validated 65-byte X9.63 point.
    /// The `signet` keystore feeds this to its ECDH-as-a-service (the SE/software
    /// backend computes the shared secret Z) so the recipient's private scalar
    /// never leaves the keystore — see [`unwrap_dek_with_shared_secret`].
    pub fn ephemeral_pubkey_x963(&self) -> Result<Vec<u8>> {
        x963_from_epk(&self.epk)
    }
}

/// Wrap a 32-byte DEK to a recipient's KEM public key (Envelope §5). No folder
/// binding (the §4.1 file envelope already binds content to `file_id` via AAD).
pub fn wrap_dek(recipient_kem_pub_x963: &[u8], dek: &[u8; 32]) -> Result<WrapEnvelope> {
    wrap_to(recipient_kem_pub_x963, dek, &[])
}

/// Unwrap a DEK wrap with the recipient's ECDH private scalar (Envelope §5).
pub fn unwrap_dek(recipient_priv_scalar: &[u8; 32], envelope: &WrapEnvelope) -> Result<[u8; 32]> {
    fixed32(unwrap_with(recipient_priv_scalar, envelope, &[])?)
}

/// Wrap a 32-byte metadata key to a recipient, bound to `root_folder_id`
/// (the 16-byte UUID) via the Concat-KDF `PartyVInfo` — P-015 (Envelope §7.2).
pub fn wrap_metadata_key(
    recipient_kem_pub_x963: &[u8],
    metadata_key: &[u8; 32],
    root_folder_id: &[u8; 16],
) -> Result<WrapEnvelope> {
    wrap_to(recipient_kem_pub_x963, metadata_key, root_folder_id)
}

/// Unwrap a metadata-key wrap. `root_folder_id` MUST match the folder the wrap
/// was created for; a mismatch derives a different KEK and AES-KW's integrity
/// check rejects it ([`CryptoError::Authentication`]) — the P-015 property.
pub fn unwrap_metadata_key(
    recipient_priv_scalar: &[u8; 32],
    envelope: &WrapEnvelope,
    root_folder_id: &[u8; 16],
) -> Result<[u8; 32]> {
    fixed32(unwrap_with(
        recipient_priv_scalar,
        envelope,
        root_folder_id,
    )?)
}

/// Unwrap a DEK wrap when the ECDH shared secret Z is supplied externally — the
/// `signet` keystore computes Z via its SE/software backend so the recipient's
/// private scalar never leaves the keystore (Envelope §5). Pair with
/// [`WrapEnvelope::ephemeral_pubkey_x963`] + the keystore's ECDH op.
pub fn unwrap_dek_with_shared_secret(z: &[u8; 32], envelope: &WrapEnvelope) -> Result<[u8; 32]> {
    fixed32(finish_unwrap(z, envelope, &[])?)
}

/// Unwrap a metadata-key wrap with an externally-supplied Z. `root_folder_id`
/// (the P-015 `party_v`) MUST match the wrap-time folder, or AES-KW rejects it.
pub fn unwrap_metadata_key_with_shared_secret(
    z: &[u8; 32],
    envelope: &WrapEnvelope,
    root_folder_id: &[u8; 16],
) -> Result<[u8; 32]> {
    fixed32(finish_unwrap(z, envelope, root_folder_id)?)
}

/// The shared wrap construction: ephemeral ECDH → Concat-KDF KEK (bound to
/// `party_v`) → AES-KW.
fn wrap_to(recipient_kem_pub_x963: &[u8], key: &[u8; 32], party_v: &[u8]) -> Result<WrapEnvelope> {
    // Validate the recipient key up front (also gives us the JWK source for the
    // `rfp` fingerprint).
    PublicKey::from_sec1_bytes(recipient_kem_pub_x963)
        .map_err(|_| CryptoError::InvalidInput("recipient kem public key"))?;

    let ephemeral = SecretKey::random(&mut OsRng);
    // §4.6 zeroization: the scalar copy, Z, and the KEK are all wiped on drop
    // (`SecretKey` itself zeroizes internally).
    let eph_scalar: Zeroizing<[u8; 32]> = Zeroizing::new(ephemeral.to_bytes().into());

    let z = Zeroizing::new(ecdh::ecdh_p256(&eph_scalar, recipient_kem_pub_x963)?);
    let kek = concatkdf::ecdh_es_a256kw_kek(&*z, party_v);
    let wrapped = keywrap::wrap(&kek, key)?; // 40 bytes for a 32-byte key

    Ok(WrapEnvelope {
        v: 1,
        alg: AlgId::EcdhEsA256Kw.as_jose().to_string(),
        epk: epk_jwk(&ephemeral.public_key()),
        ct: URL_SAFE_NO_PAD.encode(&wrapped),
        rfp: pubkey::fingerprint(recipient_kem_pub_x963)?,
    })
}

/// The shared unwrap construction (this crate computes Z from the private
/// scalar). `party_v` must match the wrap-time value.
fn unwrap_with(
    recipient_priv_scalar: &[u8; 32],
    envelope: &WrapEnvelope,
    party_v: &[u8],
) -> Result<Vec<u8>> {
    let eph_pub_x963 = x963_from_epk(&envelope.epk)?;
    let z = Zeroizing::new(ecdh::ecdh_p256(recipient_priv_scalar, &eph_pub_x963)?);
    finish_unwrap(&z, envelope, party_v)
}

/// The unwrap tail once the ECDH shared secret Z is known: check the alg, derive
/// the KEK (bound to `party_v`), AES-KW-unwrap. Shared by the scalar-based path
/// ([`unwrap_with`]) and the keystore-supplied-Z path
/// ([`unwrap_dek_with_shared_secret`] / [`unwrap_metadata_key_with_shared_secret`]).
fn finish_unwrap(z: &[u8; 32], envelope: &WrapEnvelope, party_v: &[u8]) -> Result<Vec<u8>> {
    if AlgId::from_jose(&envelope.alg)? != AlgId::EcdhEsA256Kw {
        return Err(CryptoError::UnknownAlgorithm);
    }
    let kek = concatkdf::ecdh_es_a256kw_kek(z, party_v);
    let wrapped = URL_SAFE_NO_PAD
        .decode(envelope.ct.as_bytes())
        .map_err(|_| CryptoError::InvalidInput("wrap ct base64url"))?;
    keywrap::unwrap(&kek, &wrapped)
}

/// The EC public key as the wrap envelope's `epk` JWK (uncompressed `0x04‖X‖Y`).
pub(crate) fn epk_jwk(pubkey: &PublicKey) -> EpkJwk {
    let point = pubkey.to_encoded_point(false);
    EpkJwk {
        kty: "EC".to_string(),
        crv: "P-256".to_string(),
        x: URL_SAFE_NO_PAD.encode(point.x().expect("uncompressed point has X").as_slice()),
        y: URL_SAFE_NO_PAD.encode(point.y().expect("uncompressed point has Y").as_slice()),
    }
}

/// Reconstruct + validate the 65-byte X9.63 point from an `epk` JWK.
pub(crate) fn x963_from_epk(epk: &EpkJwk) -> Result<Vec<u8>> {
    if epk.kty != "EC" || epk.crv != "P-256" {
        return Err(CryptoError::InvalidInput("epk must be EC P-256"));
    }
    let x = URL_SAFE_NO_PAD
        .decode(epk.x.as_bytes())
        .map_err(|_| CryptoError::InvalidInput("epk x base64url"))?;
    let y = URL_SAFE_NO_PAD
        .decode(epk.y.as_bytes())
        .map_err(|_| CryptoError::InvalidInput("epk y base64url"))?;
    if x.len() != 32 || y.len() != 32 {
        return Err(CryptoError::InvalidInput(
            "epk coordinates must be 32 bytes",
        ));
    }
    let mut point = Vec::with_capacity(65);
    point.push(0x04);
    point.extend_from_slice(&x);
    point.extend_from_slice(&y);
    // Reject a point that isn't actually on the curve before it reaches ECDH.
    PublicKey::from_sec1_bytes(&point)
        .map_err(|_| CryptoError::InvalidInput("epk not on curve"))?;
    Ok(point)
}

/// Convert unwrapped key material to `[u8; 32]`, wiping the source buffer in
/// both the success and the failure path (§4.6 — the buffer held key material
/// either way).
pub(crate) fn fixed32(mut bytes: Vec<u8>) -> Result<[u8; 32]> {
    let result = <[u8; 32]>::try_from(bytes.as_slice())
        .map_err(|_| CryptoError::InvalidInput("unwrapped key is not 32 bytes"));
    bytes.zeroize();
    result
}
